pinot-query-runtime/src/main/java/org/apache/pinot/query/service/dispatch/QueryDispatcher.java

```diff
@@ -189,73 +191,60 @@ int submit(long requestId, DispatchableSubPlan dispatchableSubPlan, long timeout
-                OperatorUtils.operatorStatsToJson(entry.getValue()));
-            OperatorStats operatorStats = entry.getValue();
-            ExecutionStatsAggregator rootStatsAggregator = executionStatsAggregatorMap.get(0);
-            ExecutionStatsAggregator stageStatsAggregator = executionStatsAggregatorMap.get(operatorStats.getStageId());
-            rootStatsAggregator.aggregate(null, entry.getValue().getExecutionStats(), new HashMap<>());
-            if (stageStatsAggregator != null) {
-              if (dispatchableSubPlan != null) {
-                OperatorUtils.recordTableName(operatorStats,
-                    dispatchableSubPlan.getQueryStageList().get(operatorStats.getStageId()));
-              }
-              stageStatsAggregator.aggregate(null, entry.getValue().getExecutionStats(), new HashMap<>());
-            }
+  private static void collectStats(DispatchableSubPlan dispatchableSubPlan, @Nullable OpChainStats opChainStats,
+      @Nullable Map<Integer, ExecutionStatsAggregator> executionStatsAggregatorMap) {
+    if (executionStatsAggregatorMap != null && opChainStats != null) {
+      for (Map.Entry<String, OperatorStats> entry : opChainStats.getOperatorStatsMap().entrySet()) {
+        LOGGER.info("Broker Query Execution Stats - OperatorId: {}, OperatorStats: {}", entry.getKey(),
+            OperatorUtils.operatorStatsToJson(entry.getValue()));
+        OperatorStats operatorStats = entry.getValue();
+        ExecutionStatsAggregator rootStatsAggregator = executionStatsAggregatorMap.get(0);
+        ExecutionStatsAggregator stageStatsAggregator = executionStatsAggregatorMap.get(operatorStats.getStageId());
+        rootStatsAggregator.aggregate(null, entry.getValue().getExecutionStats(), new HashMap<>());
+        if (stageStatsAggregator != null) {
+          if (dispatchableSubPlan != null) {
+            OperatorUtils.recordTableName(operatorStats,
+                dispatchableSubPlan.getQueryStageList().get(operatorStats.getStageId()));
           }
+          stageStatsAggregator.aggregate(null, entry.getValue().getExecutionStats(), new HashMap<>());
         }
-        return resultDataBlocks;
       }
-      resultDataBlocks.add(transferableBlock.getDataBlock());
     }
-    throw new RuntimeException("Timed out while receiving from mailbox: " + QueryException.EXECUTION_TIMEOUT_ERROR);
   }
 
-  private static ResultTable toResultTable(List<DataBlock> queryResult, List<Pair<Integer, String>> fields,
+  private static ResultTable toResultTable(List<TransferableBlock> queryResult, List<Pair<Integer, String>> fields,
       DataSchema sourceSchema) {
     List<Object[]> resultRows = new ArrayList<>();
     DataSchema resultSchema = toResultSchema(sourceSchema, fields);
-    for (DataBlock dataBlock : queryResult) {
+    for (TransferableBlock transferableBlock : queryResult) {
+      if (transferableBlock.isErrorBlock()) {
```

Comment:
It may make sense to do this check first before doing any DataTable creation, since otherwise we may needlessly allocate new memory and also delay inevitable failure.

```
if (queryResult.anyMatch(block -> block.isErrorBlock())) {
  throw ...
}
```

Reply:
it is impossible for pipeline breaker to create multiple success then error. (e.g. when error the result can only have 1 error block or >=0 non-error data block